File: android/src/main/kotlin/com/uqbrijesh/aws/uploader/AwsUploaderPlugin.kt
package com.uqbrijesh.aws.uploader

import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import android.content.Context
import androidx.annotation.NonNull
import com.amazonaws.auth.CognitoCachingCredentialsProvider
import com.amazonaws.mobileconnectors.s3.transferutility.*
import com.amazonaws.mobileconnectors.s3.transferutility.TransferListener
import com.amazonaws.mobileconnectors.s3.transferutility.TransferObserver
import com.amazonaws.mobileconnectors.s3.transferutility.TransferState
import com.amazonaws.mobileconnectors.s3.transferutility.TransferUtility
import com.amazonaws.mobileconnectors.s3.transferutility.TransferNetworkLossHandler
import com.amazonaws.regions.Region
import com.amazonaws.regions.Regions
import com.amazonaws.services.s3.AmazonS3Client
import com.amazonaws.services.s3.model.ObjectMetadata

import java.io.File
import java.util.concurrent.atomic.AtomicBoolean

class AwsUploaderPlugin : FlutterPlugin, MethodChannel.MethodCallHandler, EventChannel.StreamHandler {

    private lateinit var context: Context
    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var eventSink: EventChannel.EventSink? = null
    private var transferUtility: TransferUtility? = null
    private val uploads = mutableMapOf<String, TransferObserver>()

    private fun parseRegion(region: String?): Regions? {
        return try {
            if (region.isNullOrEmpty()) null else Regions.fromName(region)
        } catch (e: Exception) {
            null
        }
    }

    private fun getMimeType(filePath: String): String {
        return when (filePath.substringAfterLast('.', "").lowercase()) {
            "jpg", "jpeg" -> "image/jpeg"
            "png" -> "image/png"
            "gif" -> "image/gif"
            "webp" -> "image/webp"
            "heic", "heif" -> "image/heic"
            "pdf" -> "application/pdf"
            "mp4" -> "video/mp4"
            "mov" -> "video/quicktime"
            "avi" -> "video/x-msvideo"
            "mkv" -> "video/x-matroska"
            "mp3" -> "audio/mpeg"
            "aac" -> "audio/aac"
            "wav" -> "audio/wav"
            "ogg" -> "audio/ogg"
            "txt" -> "text/plain"
            "html", "htm" -> "text/html"
            "csv" -> "text/csv"
            "doc" -> "application/msword"
            "docx" -> "application/vnd.openxmlformats-officedocument.wordprocessingml.document"
            "xls" -> "application/vnd.ms-excel"
            "xlsx" -> "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"
            "ppt" -> "application/vnd.ms-powerpoint"
            "pptx" -> "application/vnd.openxmlformats-officedocument.presentationml.presentation"
            "zip" -> "application/zip"
            "tar" -> "application/x-tar"
            "gz" -> "application/gzip"
            else -> "application/octet-stream"
        }
    }

    override fun onAttachedToEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        methodChannel = MethodChannel(binding.binaryMessenger, "aws_uploader")
        eventChannel = EventChannel(binding.binaryMessenger, "aws_uploader_progress")
        methodChannel.setMethodCallHandler(this)
        eventChannel.setStreamHandler(this)
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: MethodChannel.Result) {
        when (call.method) {
            "startFileUpload" -> {
                val args = call.arguments as? Map<*, *>
                if (args == null) {
                    result.error("INVALID_ARGS", "Missing arguments", null)
                    return
                }
                startFileUpload(args, result)
            }
            "cancelUpload" -> {
                val args = call.arguments as? Map<*, *>
                val uploadId = args?.get("uploadId") as? String
                if (uploadId == null) {
                    result.error("INVALID_ARGS", "Missing uploadId", null)
                    return
                }
                val observer = uploads[uploadId]
                if (observer != null && transferUtility != null) {
                    transferUtility!!.cancel(observer.id)
                    uploads.remove(uploadId)
                }
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    private fun startFileUpload(args: Map<*, *>, result: MethodChannel.Result) {
        val uploadId       = args["uploadId"]       as? String ?: ""
        val bucketName     = args["bucketName"]     as? String ?: ""
        val filePath       = args["filePath"]       as? String ?: ""
        val fileName       = args["fileName"]       as? String ?: ""
        val uploadFolder   = args["uploadFolder"]   as? String ?: ""
        val regionName     = args["region"]         as? String
        val identityPoolId = args["identityPoolId"] as? String ?: ""
        val awsToken       = args["awsToken"]       as? String ?: ""
        val awsIdentityId  = args["identityId"]     as? String ?: ""
        val providerName   = args["providerName"]   as? String ?: ""
        val contentType    = args["contentType"]    as? String ?: getMimeType(filePath)

        if (uploadId.isEmpty() || bucketName.isEmpty() || filePath.isEmpty() ||
            fileName.isEmpty() || uploadFolder.isEmpty() || identityPoolId.isEmpty() ||
            awsToken.isEmpty() || awsIdentityId.isEmpty() || providerName.isEmpty()
        ) {
            result.error("INVALID_ARGS", "Missing required arguments for AWS upload", null)
            return
        }

        val file = File(filePath)
        if (!file.exists()) {
            result.error("INVALID_FILE", "File does not exist at path: $filePath", null)
            return
        }

        val region = parseRegion(regionName)
        if (region == null) {
            result.error("INVALID_REGION", "Region is null or invalid: $regionName", null)
            return
        }

        TransferNetworkLossHandler.getInstance(context)

        try {
            val developerProvider = DeveloperAuthenticationProvider(
                awsToken = awsToken,
                awsIdentityId = awsIdentityId,
                providerName = providerName,
                accountId = null,
                identityPoolId = identityPoolId,
                region = region
            )

            val credentialsProvider = CognitoCachingCredentialsProvider(
                context, developerProvider, region
            )

            val s3Client = AmazonS3Client(credentialsProvider, Region.getRegion(regionName))

            transferUtility = TransferUtility.builder()
                .context(context)
                .s3Client(s3Client)
                .build()

            val key = "$uploadFolder/$fileName"

            val metadata = ObjectMetadata()
            metadata.contentType = contentType

            val observer = transferUtility!!.upload(bucketName, key, file, metadata)
            uploads[uploadId] = observer

            // AtomicBoolean ensures only one callback (COMPLETED, FAILED, or onError)
            // calls result() even if multiple listener methods fire from different threads.
            val resultSent = AtomicBoolean(false)

            observer.setTransferListener(object : TransferListener {
                override fun onStateChanged(id: Int, state: TransferState?) {
                    when (state) {
                        TransferState.COMPLETED -> {
                            if (resultSent.compareAndSet(false, true)) {
                                val url = "https://$bucketName.s3.$regionName.amazonaws.com/$key"
                                result.success(
                                    mapOf(
                                        "uploadId" to uploadId,
                                        "url"      to url,
                                        "bucket"   to bucketName,
                                        "key"      to key,
                                        "fileName" to fileName,
                                        "filePath" to filePath,
                                    )
                                )
                                eventSink?.success(
                                    mapOf(
                                        "uploadId" to uploadId,
                                        "status"   to "completed",
                                        "progress" to 100,
                                        "url"      to url,
                                    )
                                )
                                uploads.remove(uploadId)
                            }
                        }
                        TransferState.FAILED -> {
                            if (resultSent.compareAndSet(false, true)) {
                                result.error("UPLOAD_FAILED", "Upload failed", null)
                                eventSink?.success(
                                    mapOf(
                                        "uploadId" to uploadId,
                                        "status"   to "failed",
                                        "progress" to 0,
                                    )
                                )
                                uploads.remove(uploadId)
                            }
                        }
                        else -> {}
                    }
                }

                override fun onProgressChanged(id: Int, bytesCurrent: Long, bytesTotal: Long) {
                    val progress = if (bytesTotal > 0) (bytesCurrent * 100 / bytesTotal).toInt() else 0
                    eventSink?.success(
                        mapOf(
                            "uploadId" to uploadId,
                            "status"   to "progress",
                            "progress" to progress,
                        )
                    )
                }

                override fun onError(id: Int, ex: Exception?) {
                    // Guard eventSink too — prevents a spurious "failed" event firing
                    // after a "completed" event when both COMPLETED and onError are called.
                    if (resultSent.compareAndSet(false, true)) {
                        result.error("UPLOAD_ERROR", ex?.message ?: "Unknown upload error", null)
                        eventSink?.success(
                            mapOf(
                                "uploadId" to uploadId,
                                "status"   to "failed",
                                "progress" to 0,
                                "error"    to (ex?.message ?: "Unknown error"),
                            )
                        )
                        uploads.remove(uploadId)
                    }
                }
            })

        } catch (e: Exception) {
            result.error("UPLOAD_INIT_FAILED", "Failed to initialize AWS upload: ${e.message}", null)
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
    }
}

File: ios/aws_uploader/Sources/aws_uploader/AwsUploaderPlugin.swift
import Flutter
import UIKit
import AWSCore
import AWSS3
import AWSCognito

public class AwsUploaderPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {

    private var eventSink: FlutterEventSink?
    private var uploads: [String: AWSS3TransferUtilityMultiPartUploadTask] = [:]

    public static func register(with registrar: FlutterPluginRegistrar) {
        let methodChannel = FlutterMethodChannel(name: "aws_uploader", binaryMessenger: registrar.messenger())
        let eventChannel = FlutterEventChannel(name: "aws_uploader_progress", binaryMessenger: registrar.messenger())

        let instance = AwsUploaderPlugin()
        registrar.addMethodCallDelegate(instance, channel: methodChannel)
        eventChannel.setStreamHandler(instance)
        print("✅ [AwsUploaderPlugin] Plugin registered")
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        print("📲 [AwsUploaderPlugin] Method called: \(call.method)")
        switch call.method {
        case "startFileUpload":
            guard let args = call.arguments as? [String: Any] else {
                print("❌ [AwsUploaderPlugin] Missing arguments")
                result(FlutterError(code: "INVALID_ARGS", message: "Missing arguments", details: nil))
                return
            }
            startFileUpload(args: args, result: result)
        case "cancelUpload":
            guard let args = call.arguments as? [String: Any],
                  let uploadId = args["uploadId"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing uploadId", details: nil))
                return
            }
            cancelUpload(uploadId: uploadId)
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        print("✅ [AwsUploaderPlugin] EventChannel onListen — stream subscribed")
        self.eventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        print("⚠️ [AwsUploaderPlugin] EventChannel onCancel — stream cancelled")
        self.eventSink = nil
        return nil
    }
}

// MARK: - Upload Logic
extension AwsUploaderPlugin {

    private func mimeType(for path: String) -> String {
        switch (path as NSString).pathExtension.lowercased() {
        case "jpg", "jpeg": return "image/jpeg"
        case "png":         return "image/png"
        case "gif":         return "image/gif"
        case "webp":        return "image/webp"
        case "heic", "heif": return "image/heic"
        case "pdf":         return "application/pdf"
        case "mp4":         return "video/mp4"
        case "mov":         return "video/quicktime"
        case "avi":         return "video/x-msvideo"
        case "mkv":         return "video/x-matroska"
        case "mp3":         return "audio/mpeg"
        case "aac":         return "audio/aac"
        case "wav":         return "audio/wav"
        case "ogg":         return "audio/ogg"
        case "txt":         return "text/plain"
        case "html", "htm": return "text/html"
        case "csv":         return "text/csv"
        case "doc":         return "application/msword"
        case "docx":        return "application/vnd.openxmlformats-officedocument.wordprocessingml.document"
        case "xls":         return "application/vnd.ms-excel"
        case "xlsx":        return "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"
        case "ppt":         return "application/vnd.ms-powerpoint"
        case "pptx":        return "application/vnd.openxmlformats-officedocument.presentationml.presentation"
        case "zip":         return "application/zip"
        case "gz":          return "application/gzip"
        default:            return "application/octet-stream"
        }
    }

    private func startFileUpload(args: [String: Any], result: @escaping FlutterResult) {
        print("🚀 [AwsUploaderPlugin] startFileUpload args received")

        guard
            let uploadId       = args["uploadId"]       as? String,
            let bucketName     = args["bucketName"]     as? String,
            let filePath       = args["filePath"]       as? String,
            let fileName       = args["fileName"]       as? String,
            let uploadFolder   = args["uploadFolder"]   as? String,
            let regionName     = args["region"]         as? String,
            let identityPoolId = args["identityPoolId"] as? String,
            let awsToken       = args["awsToken"]       as? String,
            let awsIdentityId  = args["identityId"]     as? String,
            let providerName   = args["providerName"]   as? String
        else {
            print("❌ [AwsUploaderPlugin] Missing required arguments")
            result(FlutterError(code: "INVALID_ARGS", message: "Missing required arguments for AWS upload", details: nil))
            return
        }

        print("✅ [AwsUploaderPlugin] Args parsed — uploadId: \(uploadId), file: \(fileName), bucket: \(bucketName)")

        let contentType = args["contentType"] as? String ?? mimeType(for: filePath)
        print("✅ [AwsUploaderPlugin] contentType: \(contentType)")

        let fileURL = URL(fileURLWithPath: filePath)
        guard FileManager.default.fileExists(atPath: fileURL.path) else {
            print("❌ [AwsUploaderPlugin] File not found at: \(filePath)")
            result(FlutterError(code: "INVALID_FILE", message: "File does not exist at path: \(filePath)", details: nil))
            return
        }
        print("✅ [AwsUploaderPlugin] File exists at: \(filePath)")

        guard AWSHelper.shared.regionType(from: regionName) != nil else {
            print("❌ [AwsUploaderPlugin] Invalid region: \(regionName)")
            result(FlutterError(code: "INVALID_REGION", message: "Region is null or invalid: \(regionName)", details: nil))
            return
        }

        AWSHelper.shared.awsToken = awsToken
        AWSHelper.shared.awsIdentityId = awsIdentityId
        print("✅ [AwsUploaderPlugin] Credentials set on AWSHelper.shared")

        let initialized = AWSHelper.shared.initializeS3(
            regionName: regionName,
            identityPoolId: identityPoolId,
            providerName: providerName
        )
        guard initialized else {
            print("❌ [AwsUploaderPlugin] initializeS3 returned false")
            result(FlutterError(code: "INVALID_REGION", message: "Provided AWS region is invalid: \(regionName)", details: nil))
            return
        }
        print("✅ [AwsUploaderPlugin] initializeS3 succeeded")

        guard let transferUtility = AWSS3TransferUtility.s3TransferUtility(forKey: "awsUploaderTransferUtility") else {
            print("❌ [AwsUploaderPlugin] TransferUtility is nil for key 'awsUploaderTransferUtility'")
            result(FlutterError(code: "UPLOAD_ERROR", message: "TransferUtility not initialized", details: nil))
            return
        }
        print("✅ [AwsUploaderPlugin] TransferUtility retrieved successfully")

        let key = "\(uploadFolder)/\(fileName)"
        print("🔧 [AwsUploaderPlugin] S3 key: \(key)")

        let expression = AWSS3TransferUtilityMultiPartUploadExpression()
        expression.progressBlock = { [weak self] _, progress in
            let pct = Int(progress.fractionCompleted * 100)
            print("📶 [AwsUploaderPlugin] Progress: \(pct)% — uploadId: \(uploadId)")
            DispatchQueue.main.async {
                self?.eventSink?([
                    "uploadId": uploadId,
                    "status": "progress",
                    "progress": pct,
                ])
            }
        }

        // Guards result() from being called more than once across both callbacks.
        var resultSent = false

        func sendResult(_ value: Any?) {
            guard !resultSent else { return }
            resultSent = true
            result(value)
        }

        func sendFailure(code: String, message: String) {
            sendResult(FlutterError(code: code, message: message, details: nil))
            self.eventSink?([
                "uploadId": uploadId,
                "status": "failed",
                "progress": 0,
                "error": message,
            ])
        }

        print("🔧 [AwsUploaderPlugin] Starting multipart upload...")
        transferUtility.uploadUsingMultiPart(
            fileURL: fileURL,
            bucket: bucketName,
            key: key,
            contentType: contentType,
            expression: expression
        ) { [weak self] _, error in
            DispatchQueue.main.async {
                if let error = error {
                    print("❌ [AwsUploaderPlugin] Upload completion error: \(error.localizedDescription)")
                    sendResult(FlutterError(code: "UPLOAD_FAILED", message: error.localizedDescription, details: nil))
                    self?.eventSink?([
                        "uploadId": uploadId,
                        "status": "failed",
                        "progress": 0,
                        "error": error.localizedDescription,
                    ])
                } else {
                    let url = "https://\(bucketName).s3.\(regionName).amazonaws.com/\(key)"
                    print("✅ [AwsUploaderPlugin] Upload completed — url: \(url)")
                    sendResult([
                        "uploadId": uploadId,
                        "url": url,
                        "bucket": bucketName,
                        "key": key,
                        "fileName": fileName,
                        "filePath": fileURL.path,
                    ])
                    self?.eventSink?([
                        "uploadId": uploadId,
                        "status": "completed",
                        "progress": 100,
                        "url": url,
                    ])
                }
                self?.uploads.removeValue(forKey: uploadId)
                print("🔧 [AwsUploaderPlugin] Upload task removed from tracking — uploadId: \(uploadId)")
            }
        }.continueWith { [weak self] task -> Any? in
            if let uploadTask = task.result {
                self?.uploads[uploadId] = uploadTask
                print("✅ [AwsUploaderPlugin] Upload task tracked — uploadId: \(uploadId)")
            } else if let error = task.error {
                print("❌ [AwsUploaderPlugin] continueWith error: \(error.localizedDescription)")
                DispatchQueue.main.async {
                    sendFailure(code: "UPLOAD_START_FAILED", message: error.localizedDescription)
                }
            } else {
                print("⚠️ [AwsUploaderPlugin] continueWith — task.result is nil")
                DispatchQueue.main.async {
                    sendFailure(code: "UPLOAD_START_FAILED", message: "Upload failed to start")
                }
            }
            return nil
        }
    }

    private func cancelUpload(uploadId: String) {
        print("🔧 [AwsUploaderPlugin] cancelUpload — uploadId: \(uploadId)")
        if let task = uploads[uploadId] {
            task.cancel()
            uploads.removeValue(forKey: uploadId)
            print("✅ [AwsUploaderPlugin] Upload cancelled")
        } else {
            print("⚠️ [AwsUploaderPlugin] cancelUpload — no task found for uploadId: \(uploadId)")
        }
    }
}
